homeassistant/components/recorder/__init__.py

```diff
@@ -0,0 +1,358 @@
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+_INSTANCE = None
+_LOGGER = logging.getLogger(__name__)
+
+# These classes will be populated during setup()
+# pylint: disable=invalid-name
+Session = None
+
+
+def query_to_states(q):
+    """Given a query object, convert all results to an array of States."""
+    import sqlalchemy.exc
+    for _ in range(0, RETRIES):
+        try:
+            return [
+                row for row in
+                (row_to_state(row) for row in q)
+                if row is not None]
+        except sqlalchemy.exc.SQLAlchemyError as e:
+            log_error(e, retry_wait=QUERY_RETRY_WAIT, rollback=True)
+    return []
+
+
+def query_to_events(q):
+    """Query the database and return a list of states."""
+    import sqlalchemy.exc
+    for _ in range(0, RETRIES):
+        try:
+            return [
+                row for row in
+                (row_to_event(row) for row in q)
+                if row is not None]
+        except sqlalchemy.exc.SQLAlchemyError as e:
+            log_error(e, retry_wait=QUERY_RETRY_WAIT, rollback=True)
+    return []
+
+
+def row_to_state(row):
+    """Convert a database row to a state."""
+    try:
+        return State(
+            row.entity_id, row.state,
+            json.loads(row.attributes),
+            dt_util.UTC.localize(row.last_changed),
+            dt_util.UTC.localize(row.last_updated)
+        )
+    except ValueError:
+        # When json.loads fails
+        _LOGGER.exception("Error converting row to state: %s", row)
+        return None
+
+
+def row_to_event(row):
+    """Convert a databse row to an event."""
+    try:
+        return Event(
+            row.event_type,
+            json.loads(row.event_data),
+            EventOrigin(row.origin),
+            dt_util.UTC.localize(row.time_fired)
+        )
+    except ValueError:
+        # When json.loads fails
+        _LOGGER.exception("Error converting row to event: %s", row)
+        return None
+
+
+def run_information(point_in_time=None):
+    """Return information about current run.
+
+    There is also the run that covers point_in_time.
+    """
+    _verify_instance()
+
+    recorder_runs = get_model('RecorderRuns')
+    if point_in_time is None or point_in_time > _INSTANCE.recording_start:
+        return recorder_runs(
+            end=None,
+            start=_INSTANCE.recording_start,
+            closed_incorrect=False)
+
+    return query('RecorderRuns').filter(
+        (recorder_runs.start < point_in_time) &
+        (recorder_runs.end > point_in_time)).first()
+
+
+def setup(hass, config):
+    """Setup the recorder."""
+    # pylint: disable=global-statement
+    # pylint: disable=too-many-locals
+    global _INSTANCE
+    purge_days = config.get(DOMAIN, {}).get(CONF_PURGE_DAYS)
+
+    db_url = config.get(DOMAIN, {}).get(CONF_DB_URL, None)
+    if not db_url:
+        db_url = DEFAULT_URL.format(
+            hass_config_path=hass.config.path(DEFAULT_DB_FILE))
+
```

Comment:
Running the migration automatically could take a very long time, may run us out of disk space, or may wear out SD cards.  We could migrate the current day's data automatically, though?
With regards to moving the script into `homeassistant.scripts`, that's easy and lets us adapt the conversion script for automation if needed.